### ISSUE TITLE
Callbacks in examples are not called

### ISSUE THREAD

**Issue by stacywile** [spectcl/spectcl#97]
For example, in echo.js, the console.log at line 18 doesn't happen

**Comment by ryanbmilbourne** [comment]
Can you please post an example with `debug` output enabled and your code that is calling `spectcl`?

**Comment by ryanbmilbourne** [comment]
```
DEBUG=spectcl ./echo.js                                                                                                                                                                                                 
  spectcl [spawn] command:      echo +0ms
  spectcl [spawn] cmdParams:    [ 'hello' ] +5ms
  spectcl [spawn] cmdOptions:   {} +3ms
  spectcl [spawn] spawnOptions: {} +0ms
  spectcl [spawn] created cacheStream +0ms
  spectcl [spawn] child tty:    /dev/pts/11 +8ms
  spectcl [spawn] child pid:    7002 +1ms
  spectcl [expect] [7002] expectations: /hello/ +0ms
  spectcl [expect] [7002] does input match on RegExp pattern "/hello/"? no +1ms
  spectcl [expect] [7002] read eof +6ms
  spectcl [expect] [7002] warning: no callback defined for EXP_EOFὊ5 +1ms
```

**Comment by ryanbmilbourne** [comment]
@stacywile on what platform are you running the spectcl example?
I can replicate the issue on `CentOS Linux release 7.1.1503 (Core)`.

**Comment by ryanbmilbourne** [comment]
A small wrinkle.  I tried another example, `node.js`.  The output came through just fine:
```
DEBUG=spectcl ./node.js 
  spectcl [spawn] command:      node +0ms
  spectcl [spawn] cmdParams:    [ '--interactive' ] +6ms
  spectcl [spawn] cmdOptions:   undefined +2ms
  spectcl [spawn] spawnOptions: {} +0ms
  spectcl [spawn] created cacheStream +1ms
  spectcl [spawn] child tty:    /dev/pts/1 +11ms
  spectcl [spawn] child pid:    9460 +0ms
  spectcl [expect] [9460] expectations: />/ +1ms
  spectcl [expect] [9460] does input match on RegExp pattern "/>/"? no +1ms
  spectcl [expect] [9460] does input match on RegExp pattern "/>/"? no +102ms
  spectcl [expect] [9460] does input match on RegExp pattern "/>/"? yes +2ms
  spectcl [expect] [9460] found match in data +1ms
  spectcl [expect] [9460] calling handler for />/ +0ms
  spectcl [send] writing to child stdin +0ms
  spectcl [expect] [9460] expectations: > +1ms
  spectcl [expect] [9460] does input match on string ">"? no +0ms
  spectcl [send] wrote to child stdin +1ms
  spectcl [expect] [9460] does input match on string ">"? no +9ms
  spectcl [expect] [9460] does input match on string ">"? yes +1ms
  spectcl [expect] [9460] found match in data +1ms
  spectcl [expect] [9460] calling handler for > +0ms
  spectcl [send] writing to child stdin +0ms
output was:
  console.log('testing')
testing
undefined
>
all done
  spectcl [send] wrote to child stdin +2ms
  spectcl [spawn] child 9460 exit. code: 0      signal: null +4ms
```
The output matches what my Mac (which is not demonstrating the issue) is outputting.

I'll try some others.

**Comment by stacywile** [comment]
On CentOS Linux release 7.0.1406 (Core), node version 0.12.7
[swile@samserv02.sb.viasat.com ~/spec]$ examples/echo.js
  spectcl [spawn] command:      echo +0ms
  spectcl [spawn] cmdParams:    [ 'hello' ] +14ms
  spectcl [spawn] cmdOptions:   {} +5ms
  spectcl [spawn] spawnOptions: {} +0ms
  spectcl [spawn] created cacheStream +1ms
  spectcl [spawn] child tty:    /dev/pts/20 +19ms
  spectcl [spawn] child pid:    7729 +0ms
  spectcl [expect] [7729] expectations: /hello/ +2ms
  spectcl [expect] self.cache =  +1ms
  spectcl [expect] [7729] does input match on RegExp pattern "/hello/"? no +0ms
  spectcl [expect] [7729] read eof +11ms
  spectcl [expect] [7729] warning: no callback defined for EXP_EOFὊ5 +1ms
all done
[swile@samserv02.sb.viasat.com ~/spec]$

**Comment by stacywile** [comment]
What's strange is the same example in the mocha test works fine

**Comment by gcochard** [comment]
The mocha test currently mocks `child_process` so the behavior might be slightly different due to that. I don't see how this can be happening, as the `EOF` is read (which means the process exited and all output is done). Can one of you try modifying the example to pass `{noPty:true}` as the `spawnOptions` and see if that changes anything? You might have to run it a few times to catch it.

The biggest hint here to me is that programs which do not accept any user input on stdin (such as echo) tend to demonstrate this erroneous behavior. If we switch all of our examples to programs that read user input, I think we'll solve this mystery. There's no real need to use `spectcl` to match output of a program that does not accept input, and that isn't really a goal of our project. You can simply use `child_process.exec` directly for that.

**Comment by ryanbmilbourne** [comment]
As best I can tell, no data is being detected and added to the cache.  `onData` isn't being called on the my CentOS box.  I can replicate this issue with the `echo` and `ls` options.

Running without the PTY wrapper resolves the issue so this could be rooted in an issue with `child_pty`, or our use of it.  I'll investigate further and open issues with `child_pty` if needed.

**Comment by ryanbmilbourne** [comment]
I've removed `spectcl` from the equation and setup a basic child spawn that mirrors the way that spectcl does it:
``` javascript
'use strict'
var pty = require('child_pty')

child = pty.spawn('echo', ['hello'])
child.stdout.pipe(process.stdout);
```
This code outputs "hello" on my mac, but not on my CentOS box, which mirrors what we're seeing above.  Based on that, we're either spawning in a sub-optimal way (my guess) or there's a bug in `child_pty`.  My first guess is that we're experiencing a corner case where the output is being swallowed before the pipe instruction is completed.

**Comment by gcochard** [comment]
What about this:

```js
'use strict'
var pty = require('child_pty')
pty.spawn('echo', ['hello']).stdout.pipe(process.stdout)
```

**Comment by ryanbmilbourne** [comment]
I just tried that and using `options.stdio` to redirect.  The issue with the redirect is that by default all of the child's descriptors are piped to the `pty` wrapper (`stdio: ['pty', 'pty','pty']`), and modifying it at spawn-time causes a crash.
Chaining the call as above made no impact.

**Comment by ryanbmilbourne** [comment]
Adding a `stdout` listener through chaining also does not resolve the issue:
```
pty.spawn('echo', ['hello']).stdout.on('data', function(data){
    console.log(data.toString());
});
```
The above block will output the buffered data on my Mac but not on my CentOS dev box.  I also tried using the `pty` stream instead of `stdout`.

It's worth noting that `child_pty`'s tests pass on my CentOS box.  The tests spawn and use a Node.js tool (which needs a bit of time to spin up its interpretive environment), not a core tool like `echo` or `ls`, so it could be we've found a corner case that the maintainer hasn't caught.
